frontend/test/resolution/testManage.cpp

```diff
@@ -0,0 +1,312 @@
+      foundEnter = true;
+      assert(aa.fn()->id().symbolPath() == "M.R.enterContext");
+    } else if (aa.action() == AssociatedAction::EXIT_CONTEXT) {
+      foundExit = true;
+      assert(aa.fn()->id().symbolPath() == "M.R.exitContext");
+    }
+  }
+  assert(foundEnter && foundExit);
+
+  auto byRef = findVariable(modules[0], "byRef");
+  auto byRefParent = parsing::parentAst(ctx, byRef)->toAs()->symbol();
+  assert(byRefParent);
+  assert(rr.hasAst(byRefParent));
+  foundEnter = foundExit = false;
+  for (auto aa : rr.byAst(byRefParent).associatedActions()) {
+    if (aa.action() == AssociatedAction::ENTER_CONTEXT) {
+      foundEnter = true;
+      assert(aa.fn()->id().symbolPath() == "M.R.enterContext#1");
+    } else if (aa.action() == AssociatedAction::EXIT_CONTEXT) {
+      foundExit = true;
+      assert(aa.fn()->id().symbolPath() == "M.R.exitContext");
+    }
+  }
+  assert(foundEnter && foundExit);
+
+  auto byConstRef = findVariable(modules[0], "byConstRef");
+  auto byConstRefParent = parsing::parentAst(ctx, byConstRef)->toAs()->symbol();
+  assert(byConstRefParent);
+  assert(rr.hasAst(byConstRefParent));
+  foundEnter = foundExit = false;
+  for (auto aa : rr.byAst(byConstRefParent).associatedActions()) {
+    if (aa.action() == AssociatedAction::ENTER_CONTEXT) {
+      foundEnter = true;
+      assert(aa.fn()->id().symbolPath() == "M.R.enterContext#2");
+    } else if (aa.action() == AssociatedAction::EXIT_CONTEXT) {
+      foundExit = true;
+      assert(aa.fn()->id().symbolPath() == "M.R.exitContext");
+    }
+  }
+  assert(foundEnter && foundExit);
+}
+
+static void testNoExplicitImplements() {
+  // We complain about not having an explicit ': contextManager', but still resolve.
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    record R {
+      proc enterContext() {
+        return 42;
+      }
+      proc exitContext(in error: owned Error?) {}
+    }
+    manage new R() as res {}
+    )""";
+
+  auto x = resolveTypesOfVariables(ctx, program, {"res"}).at("res");
+  assert(x.type()->isIntType());
+}
+
+static void testWithoutInterfaceMatch() {
+  // Unlike in production, we like the 'owned Error' formal to be called 'error',
+  // and not anything else. This is a consequence of leaning on interfaces
+  // for context managers.
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    record R : contextManager {
+      proc enterContext() {
+        return 42;
+      }
+      proc exitContext(in err: owned Error?) {} /* wrong name of 'err' */
+    }
+    manage new R() as res {}
+    )""";
+
+  auto x = resolveTypesOfVariables(ctx, program, {"res"}).at("res");
+  assert(x.isUnknownOrErroneous());
+
+  bool foundError = false;
+  for (auto& error : guard.errors()) {
+    if (error->type() == ErrorType::InterfaceMissingFn) {
+      foundError = true;
+      break;
+    }
+  }
+  assert(foundError);
+  assert(guard.realizeErrors());
```

Comment:
Technically this prevents failure if there are any other errors, right? Is that intentional?

Reply:
Well, it was intentional (I don't like tests breaking due to the the number of errors changing), but in this case the program is simple enough that I'll just bite the bullet.

Reply:
Fair enough. I worry about masking new errors that might be meaningful, though in practice it'd get hit somewhere else.